File: 238. Product of Array Except Self.py
# 238. Product of Array Except Self


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [1] * n
        
        left_product = 1
        for i in range(n):
            ans[i] = left_product
            left_product *= nums[i]

        right_product = 1
        for i in range(n-1, -1, -1):
            ans[i] *= right_product
            right_product *= nums[i]
        
        return ans







        # check = nums.count(0)
        # if check >= 2:
        #     return [0] * len(nums)

        # mul = 1
        # for n in nums:
        #     if n != 0:
        #         mul *= n

        # ans = []
        # for n in nums:
        #     if check == 0:
        #         ans.append(mul // n)
        #     elif check == 1:
        #         if n != 0:
        #             ans.append(0)
        #         else:
        #             ans.append(mul)
        # return ans

File: 238. Product of Array Except Self.cpp
// 238. Product of Array Except Self

#include <vector>

class Solution {
    public:
        std::vector<int> productExceptSelf(std::vector<int>& nums) {
            int prefix {1};
            int n = nums.size();
            std::vector<int> ans(nums.size(), 1);
    
            for (int i = 0; i < n; i++) {
                ans[i] *= prefix;
                prefix *= nums[i];
                // i++; 
            }
    
            int suffix {1};
            for (int i = n-1; i > -1; i--) {
                ans[i] *= suffix;
                suffix *= nums[i];
                // i++; 
            }
            return ans;
        }
    };